### PR TITLE
Sync src/iostream.c and hpcgap/src/iostream.c

### PR DESCRIPTION
**Pull request by fingolfin**
This PR gets rid of `hpcgap/src/iostream.c` by fully syncing it with `src/iostream.c`. It also adds some comments (including FIXMEs) and attempts to resolve some potential race conditions in the iostream code when using HPC-GAP.

### CODE REVIEW

File: src/iostream.c [hunk 1]
```diff
@@ -531,15 +551,23 @@ Obj FuncREAD_IOSTREAM( Obj self, Obj stream, Obj len )
...
+  }
+  
   /* HandleChildStatusChanges(pty);   Omit this to allow picking up "trailing" bytes*/
   ret = ReadFromPty2(pty, CSTR_STRING(string), INT_INTOBJ(len), 1);
-  if (ret == -1)
+  if (ret == -1) {
+    HashUnlock(PtyIOStreams);
     return Fail;
+  }
   SET_LEN_STRING(string, ret);
```
Comment (ChrisJefferson) [inline src/iostream.c]: You can't call `ResizeBag` with a `HashLock` held -- but in this case it should be safe to pull the lock back to here.

File: src/iostream.c [hunk 2]
```diff
@@ -549,38 +577,57 @@ Obj FuncREAD_IOSTREAM_NOWAIT(Obj self, Obj stream, Obj len)
...
   ret = ReadFromPty2(pty, CSTR_STRING(string), INT_INTOBJ(len), 0);
-  if (ret == -1)
+  if (ret == -1) {
+    HashUnlock(PtyIOStreams);
     return Fail;
+  }
   SET_LEN_STRING(string, ret);
   ResizeBag(string, SIZEBAG_STRINGLEN(ret));
+
+  HashUnlock(PtyIOStreams);
```
Comment (ChrisJefferson) [inline src/iostream.c]: Again here, pull HashUnlock further back.

File: src/iostream.c (hunk 1 above)
Comment (fingolfin) [inline src/iostream.c]: OK -- so, were is this documented (and out of curiosity: why is it so?). This strikes me as very important information (sadly, most HPC-GAP kernel code seems to undocumented, with no hints about requirements or prerequisites or gotachas or about what they do... :/).

Comment (ChrisJefferson) [inline src/iostream.c]: Well actually, you mustn't call anything which might take a lock of any kind while a `HashLock` is held, including another `HashLock` -- but that might not be documented.

File: src/iostream.c (hunk 2 above)
Comment (fingolfin) [inline src/iostream.c]: Fixed

File: src/iostream.c [hunk 3]
```diff
@@ -359,48 +378,44 @@ Int StartChildProcess ( Char *dir, Char *prg, Char *args[] )
...
+    HashUnlock(PtyIOStreams);
     return -1;
 }
 
 
-void HandleChildStatusChanges( UInt pty)
+// This function assues that the caller invoked HashLock(PtyIOStreams).
+// It unlocks just before throwing any error.
+static void HandleChildStatusChanges( UInt pty)
 {
```
Comment (markuspf) [inline src/iostream.c]: again a superflous space (or a missing one) ;)

File: src/iostream.c [hunk 4]
```diff
@@ -434,9 +449,7 @@ Obj FuncCREATE_PTY_IOSTREAM( Obj self, Obj dir, Obj prog, Obj args )
 }
 
 
-
-
-Int ReadFromPty2( UInt stream, Char *buf, Int maxlen, UInt block)
+static Int ReadFromPty2( UInt stream, Char *buf, Int maxlen, UInt block)
```
Comment (markuspf) [inline src/iostream.c]: Spaces here are also inconsistent

File: src/iostream.c [hunk 5]
```diff
@@ -359,48 +378,44 @@ Int StartChildProcess ( Char *dir, Char *prg, Char *args[] )
...
     PtyIOStreams[stream].inuse = 0;
     FreeStream(stream);
+    HashUnlock(PtyIOStreams);
     return -1;
 }
 
 
-void HandleChildStatusChanges( UInt pty)
+// This function assues that the caller invoked HashLock(PtyIOStreams).
+// It unlocks just before throwing any error.
```
Comment (markuspf) [inline src/iostream.c]: assumes

File: src/iostream.c [hunk 6]
```diff
@@ -480,17 +493,19 @@ Int ReadFromPty2( UInt stream, Char *buf, Int maxlen, UInt block)
...
+static UInt WriteToPty ( UInt stream, Char *buf, Int len )
 {
     Int         res;
     Int         old;
-/*  struct timeval tv; */
-/*  fd_set      writefds; */
-/*  int retval; */
-    if (len < 0)
+    if (len < 0) {
+      // FIXME: why do allow 'len' to be negative here? To allow
```
Comment (markuspf) [inline src/iostream.c]: Either "do we" or delete the do?

File: src/iostream.c [hunk 7]
```diff
@@ -480,17 +493,19 @@ Int ReadFromPty2( UInt stream, Char *buf, Int maxlen, UInt block)
...
+      // invoking a "raw" version of `write` perhaps? But we don't
+      // seem to use that anywhere. So perhaps get rid of it or
+      // even turn it into an error?!
```
Comment (markuspf) [inline src/iostream.c]: I think allowing negative length is obscure and obviously undocumented, so I am all for turning this into an error.

File: src/iostream.c [hunk 8]
```diff
@@ -94,63 +98,62 @@ typedef struct {
...
   {
       stream = FreePtyIOStreams;
       FreePtyIOStreams = PtyIOStreams[stream].childPID;
   }
-  HashUnlock(PtyIOStreams);
   return stream;
 }
 
-void FreeStream( UInt stream)
+static void FreeStream( UInt stream)
```
Comment (markuspf) [inline src/iostream.c]: This either has one space  too many, or one too few.

File: src/iostream.c (hunk 7 above)
Comment (fingolfin) [inline src/iostream.c]: I agree; do you want me to add that to this PR then?

**Comment by fingolfin** [comment]
@ChrisJefferson so... anything left to address?

**Comment by ChrisJefferson** [comment]
No, I'm happy with this being merged.

**Comment by fingolfin** [comment]
@markuspf if you are also OK with it, an we merge it now? I am afraid it'll bitrot if/when master is updated.

**Pull request merged**